rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/KeyPairProducer.java

```diff
@@ -0,0 +1,136 @@
+ */
+package org.ow2.proactive.resourcemanager.utils;
+
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.log4j.Logger;
+
+
+public class KeyPairProducer {
+
+    private static final Logger LOGGER = Logger.getLogger(KeyPairProducer.class);
+
+    // should be aligned with time unit used in Object::wait()
+    private static final TimeUnit timeUnit = TimeUnit.MILLISECONDS;
+
+    private AtomicInteger numberOfWorkers;
+
+    private final Future<?> future;
+
+    private BlockingQueue<KeyPair> keyPairs = new LinkedBlockingQueue<>();
+
+    private long loopTimeout = 1000;
+
+    public KeyPairProducer(int numberOfworkers) {
+        this.numberOfWorkers = new AtomicInteger(numberOfworkers);
+        final ExecutorService executorService = Executors.newSingleThreadExecutor();
+        future = executorService.submit(new Runnable() {
+            @Override
+            public void run() {
+                while (!Thread.interrupted()) {
+                    try {
+                        // if we obeserve less than 'numberOfWorkers' elements than we going to
+                        // add 'numberOfWorkers' more
+                        int numberOfKeysToAdd = numberOfWorkers.get() - keyPairs.size();
+                        if (numberOfKeysToAdd > 0) {
+                            for (int i = 0; i < numberOfKeysToAdd; ++i) {
+                                keyPairs.put(generateAndAddKeyPairs());
+                            }
+                        }
+
+                        // sleep 1 second, ot until notification (from getKeyPair method)
+                        synchronized (keyPairs) {
+                            keyPairs.wait(loopTimeout);
+                        }
+                    } catch (InterruptedException e) {
+                        LOGGER.error("Thread was interrupted.", e);
+                        Thread.currentThread().interrupt();
+                    } catch (Throwable e) {
+                        // ignore all exception except InterruptedException
+                        LOGGER.error(e);
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * @return keyPair from head of the keyPairs queue
+     */
+    public KeyPair getKeyPair() throws NoSuchAlgorithmException {
+        // if by change queue is empty then we notify background thread which apparently sleeps
+        // so that we do not wait whole timeout.
+        if (keyPairs.isEmpty()) {
+            synchronized (keyPairs) {
+                keyPairs.notify();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') "notify" may not wake up the appropriate thread. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2446)

Reply:
Instead of using wait/notify mechanism, I still believe that it's easier to simply generate the key inside this method if the queue is empty.